storage/innobase/os/os0proc.cc

```diff
@@ -64,44 +61,43 @@ void*
 os_mem_alloc_large(
 	ulint*	n)
 {
-	void*	ptr;
+	void*	ptr = NULL;
 	ulint	size;
 #if defined HAVE_LINUX_LARGE_PAGES && defined UNIV_LINUX
-	int shmid;
-	struct shmid_ds buf;
+	int     mapflag;
+	int i = 0;
+	size_t large_page_size, adjusted_size;
 
-	if (!os_use_large_pages || !os_large_page_size) {
+	if (!my_use_large_pages) {
 		goto skip;
 	}
 
-	/* Align block size to os_large_page_size */
-	ut_ad(ut_is_2pow(os_large_page_size));
-	size = ut_2pow_round(*n + (os_large_page_size - 1),
-			     os_large_page_size);
+	/* MAP_HUGE_SHIFT added linux-3.8. Take largest HUGEPAGE size */
+	while ((large_page_size = my_next_large_page_size(*n, &i))) {
+		mapflag = MAP_PRIVATE | MAP_ANONYMOUS | MAP_HUGETLB | my_bit_log2(large_page_size) << MAP_HUGE_SHIFT;
```

Comment:
`my_bit_log2()` takes `ulong`, not `size_t`.
Do we really need both the logarithm and bit-shifting? Could we just reject not-power-of-2 page sizes (which are extremely unlikely in practice) when parsing the data from the kernel? See `ut_is_2pow()`.

Reply:
> Do we really need both the logarithm and bit-shifting?

Yes. log2(2M) is 20 and there's no way to get from 2M to 20 * 2^^26 without both transforms. Added size_t equivalent of `my_bit_log2`